src/main/java/org/jenkins/tools/test/PluginCompatTester.java

```diff
@@ -417,14 +412,30 @@ private void testPluginAgainst(
         forExecutionHooks.put("coreCoordinates", coreCoordinates);
         forExecutionHooks.put("config", config);
         forExecutionHooks.put("pluginDir", pluginCheckoutDir);
-        forExecutionHooks.put("types", testTypes);
         pcth.runBeforeExecution(forExecutionHooks);
         args = (List<String>) forExecutionHooks.get("args");
-        Set<String> types = new HashSet<>((List<String>) forExecutionHooks.get("types"));
-        userProperties.put("types", String.join(",", types));
+
+        Map<String, String> properties = new LinkedHashMap<>(config.getMavenProperties());
+        properties.put("overrideWar", config.getWar().toString());
+        properties.put("jenkins.version", coreCoordinates.version);
+        properties.put("useUpperBounds", "true");
+        if (new VersionNumber(coreCoordinates.version).isOlderThan(new VersionNumber("2.382"))) {
+            /*
+             * Versions of Jenkins prior to 2.382 are susceptible to JENKINS-68696, in which
+             * javax.servlet:servlet-api comes from core at version 0. This is an intentional trick
+             * to prevent this library from being used, and we do not want it to be upgraded to a
+             * nonzero version (which is not a realistic test scenario) just because it happens to
+             * be on the class path of some plugin and triggers an upper bounds violation.
+             */
+            properties.put("upperBoundsExcludes", "javax.servlet:servlet-api");
+        }
 
         // Execute with tests
-        runner.run(mconfig, pluginCheckoutDir, buildLogFile, args.toArray(new String[0]));
+        runner.run(
+                Collections.unmodifiableMap(properties),
+                pluginCheckoutDir,
+                buildLogFile,
+                args.toArray(new String[0]));
```

Comment:
This is the only invocation of Maven where we are running tests (as opposed to compilation), so we set some things that are specific to PCT runs: `overrideWar`, `jenkins.version`, and `useUpperBounds`. We also work around JENKINS-68696 if we have to. This frees consumers of this burden.